packages/core/src/tracing/instrumentation/messaging/nats.js

```diff
@@ -19,39 +19,116 @@ exports.init = function init() {
+
+        // nats.requestOne uses nats.request internally, so there is no need to instrument requestOne separately
+        shimmer.wrap(client.constructor.prototype, 'request', shimRequest.bind(null, client.options.url));
 
-      shimmer.wrap(client.constructor.prototype, 'subscribe', shimSubscribe.bind(null, client.options.url));
-      clientHasBeenInstrumented = true;
+        shimmer.wrap(client.constructor.prototype, 'subscribe', shimSubscribe.bind(null, client.options.url, true));
+        clientHasBeenInstrumented = true;
+      }
     }
+
     return client;
   };
 }
 
-function shimPublish(natsUrl, originalFunction) {
+function shimPublish(natsUrl, isLatest, originalFunction) {
   return function () {
     const originalArgs = new Array(arguments.length);
     for (let i = 0; i < arguments.length; i++) {
       originalArgs[i] = arguments[i];
     }
 
-    return instrumentedPublish(this, originalFunction, originalArgs, natsUrl);
+    return instrumentedPublish(this, originalFunction, originalArgs, natsUrl, isLatest);
   };
 }
 
-function instrumentedPublish(ctx, originalPublish, originalArgs, natsUrl) {
+function addSuppressionHeaders(args, isLatest) {
+  if (!isLatest) return;
+  if (!natsModule || !natsModule.headers) return;
+  if (!connectionObject || !connectionObject.info) return;
+  // not every docker nats version supports headers
+  if (!connectionObject.info.headers) return;
+
+  const opts = args[2];
+  const Headers = natsModule.headers;
+
+  if (opts && opts.headers && typeof opts.headers === 'object') {
+    opts.headers.append(constants.traceLevelHeaderName, '0');
+  } else if (!opts) {
+    const h = Headers();
+    h.append(constants.traceLevelHeaderName, '0');
+    args[2] = { headers: h };
+  } else {
+    const h = Headers();
+    h.append(constants.traceLevelHeaderName, '0');
+    opts.headers = h;
+  }
+}
+
+function addTraceCorrelationHeaders(args, isLatest, span) {
+  if (!isLatest) return;
+  if (!natsModule || !natsModule.headers) return;
+  if (!connectionObject || !connectionObject.info) return;
+  // not every docker nats version supports headers
+  if (!connectionObject.info.headers) return;
+
+  const opts = args[2];
+  const Headers = natsModule.headers;
+
+  if (opts && opts.headers && typeof opts.headers === 'object') {
+    opts.headers.append(constants.traceIdHeaderName, span.t);
+    opts.headers.append(constants.spanIdHeaderName, span.s);
+  } else if (!opts) {
+    const h = Headers();
+    h.append(constants.traceIdHeaderName, span.t);
+    h.append(constants.spanIdHeaderName, span.s);
+    args[2] = { headers: h };
+  } else {
+    const h = Headers();
+    h.append(constants.traceIdHeaderName, span.t);
+    h.append(constants.spanIdHeaderName, span.s);
+    opts.headers = h;
+  }
+}
+
+function instrumentedPublish(ctx, originalPublish, originalArgs, natsUrl, isLatest) {
   if (cls.skipExitTracing({ isActive })) {
+    addSuppressionHeaders(originalArgs, isLatest);
```

Comment:
The use case is skipping the tracing because the collector has not yet connected to the agent. The downstream service might also be instrumented and already connected to an agent, in this case the expected behavior is that the downstream service starts a trace with a root span. If we send X-INSTANA-L=0 downstream incorrectly, the downstream tracer would suppress instead of starting a trace.